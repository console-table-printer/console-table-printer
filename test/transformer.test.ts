import { Table } from '../index';
import {
  ValueTransformer,
  CellValue,
  ComplexOptions,
} from '../src/models/external-table';

describe('Table Data Transformers Tests', () => {
  it('should render transformed table correctly', () => {
    const testData = [
      { firstName: 'John', lastName: 'Doe', salary: 50000 },
      { firstName: 'Jane', lastName: 'Smith', salary: 60000 },
    ];
    const tranformer: ValueTransformer = (data: CellValue): CellValue =>
      Number(data).toFixed(2);

    const options = {
      columns: [
        { name: 'firstName' },
        { name: 'lastName' },
        { name: 'salary', transformer: tranformer },
      ],
      computedColumns: [
        {
          name: 'fullName',
          title: 'Full Name',
          function: (row) => `${row.firstName} ${row.lastName}`,
        },
        {
          name: 'taxed_salary',
          title: 'After Tax',
          function: (row) => (row.salary * 0.8).toFixed(2),
        },
      ],
    } as ComplexOptions;

    const table = new Table(options);

    table.addRows(testData);

    const rendered = table.render();
    expect(rendered).toMatchSnapshot();
    table.printTable();
  });

  it('should not transform data when print multiple times', () => {
    const testData = [
      { firstName: 'John', lastName: 'Doe', salary: 50000 },
      { firstName: 'Jane', lastName: 'Smith', salary: 60000 },
    ];
    const tranformer: ValueTransformer = (data: CellValue): CellValue =>
      Number(data).toFixed(2);

    const options = {
      columns: [
        { name: 'firstName' },
        { name: 'lastName' },
        { name: 'salary', transformer: tranformer },
      ],
      computedColumns: [
        {
          name: 'fullName',
          title: 'Full Name',
          function: (row) => `${row.firstName} ${row.lastName}`,
        },
        {
          name: 'taxed_salary',
          title: 'After Tax',
          function: (row) => (row.salary * 0.8).toFixed(2),
        },
      ],
    } as ComplexOptions;

    const table = new Table(options);

    table.addRows(testData);

    const firstRendered = table.render();
    const secondRendered = table.render();
    expect(firstRendered).toMatchSnapshot();
    expect(secondRendered).toMatchSnapshot();
    expect(firstRendered).toEqual(secondRendered);
    table.printTable();
  });
});
