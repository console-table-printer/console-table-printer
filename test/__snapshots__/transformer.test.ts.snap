// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Data Transformers Tests should not transform data when print multiple times 1`] = `
"┌───────────┬──────────┬──────────┬────────────┬───────────┐
│[37m [0m[01mfirstName[0m │[37m [0m[01mlastName[0m │[37m [0m[01m  salary[0m │[37m [0m[01m Full Name[0m │[37m [0m[01mAfter Tax[0m │
├───────────┼──────────┼──────────┼────────────┼───────────┤
│[37m [0m[37m     John[0m │[37m [0m[37m     Doe[0m │[37m [0m[37m50000.00[0m │[37m [0m[37m  John Doe[0m │[37m [0m[37m 40000.00[0m │
│[37m [0m[37m     Jane[0m │[37m [0m[37m   Smith[0m │[37m [0m[37m60000.00[0m │[37m [0m[37mJane Smith[0m │[37m [0m[37m 48000.00[0m │
└───────────┴──────────┴──────────┴────────────┴───────────┘"
`;

exports[`Table Data Transformers Tests should not transform data when print multiple times 2`] = `
"┌───────────┬──────────┬──────────┬────────────┬───────────┐
│[37m [0m[01mfirstName[0m │[37m [0m[01mlastName[0m │[37m [0m[01m  salary[0m │[37m [0m[01m Full Name[0m │[37m [0m[01mAfter Tax[0m │
├───────────┼──────────┼──────────┼────────────┼───────────┤
│[37m [0m[37m     John[0m │[37m [0m[37m     Doe[0m │[37m [0m[37m50000.00[0m │[37m [0m[37m  John Doe[0m │[37m [0m[37m 40000.00[0m │
│[37m [0m[37m     Jane[0m │[37m [0m[37m   Smith[0m │[37m [0m[37m60000.00[0m │[37m [0m[37mJane Smith[0m │[37m [0m[37m 48000.00[0m │
└───────────┴──────────┴──────────┴────────────┴───────────┘"
`;

exports[`Table Data Transformers Tests should render transformed table correctly 1`] = `
"┌───────────┬──────────┬──────────┬────────────┬───────────┐
│[37m [0m[01mfirstName[0m │[37m [0m[01mlastName[0m │[37m [0m[01m  salary[0m │[37m [0m[01m Full Name[0m │[37m [0m[01mAfter Tax[0m │
├───────────┼──────────┼──────────┼────────────┼───────────┤
│[37m [0m[37m     John[0m │[37m [0m[37m     Doe[0m │[37m [0m[37m50000.00[0m │[37m [0m[37m  John Doe[0m │[37m [0m[37m 40000.00[0m │
│[37m [0m[37m     Jane[0m │[37m [0m[37m   Smith[0m │[37m [0m[37m60000.00[0m │[37m [0m[37mJane Smith[0m │[37m [0m[37m 48000.00[0m │
└───────────┴──────────┴──────────┴────────────┴───────────┘"
`;
